File: android/app/src/main/java/com/floattime/app/FloatTimeService.java
package com.floattime.app;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import org.json.JSONObject;

import java.net.HttpURLConnection;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * 悬浮时间前台服务
 * 使用通知栏显示实时校准时间
 *
 * 优化:
 * - 通知 Builder 复用: 不每次 new Builder，只更新 title/text
 * - 夜间模式逻辑抽成工具方法避免重复
 * - Theme CHANGE_THEME action 处理
 * - SharedPreferences 写入节流
 * - 连接超时提升到 5s
 * - 统一使用 LiveUpdateManager 的通知渠道
 */
public class FloatTimeService extends Service {

    private static final String TAG = "FloatTimeService";
    private static final String PREFS_NAME = "FloatTimePrefs";
    private static final String KEY_THEME_MODE = "theme_mode";
    private static final String KEY_OFFSET_MS = "offset_ms";
    private static final String KEY_TIME_SOURCE = "time_source";

    private static final int NOTIFICATION_ID = 20240320;

    // 优化: 通知更新间隔从 100ms 提升到 200ms
    // 原来每 100ms 都重建完整通知（Builder + JSON + 反射），200ms 足够人类视觉感知
    private static final int UPDATE_INTERVAL_MS = 200;

    private static final int CONNECT_TIMEOUT_MS = 5000; // 修复: 3000ms 对淘宝 API 太紧
    private static final int READ_TIMEOUT_MS = 5000;

    private static final SimpleDateFormat TIME_FORMAT =
            new SimpleDateFormat("HH:mm:ss", Locale.getDefault());

    private static volatile boolean sIsRunning = false;
    public static boolean isRunning() { return sIsRunning; }

    private NotificationManager mNotificationManager;
    private LiveUpdateManager mLiveUpdateManager;

    private SharedPreferences mPrefs;
    private Handler mHandler;
    private Runnable mTimeRunnable;

    private String mTimeSource = "taobao";
    private long mOffsetMs = 0;
    private final AtomicBoolean mIsSyncing = new AtomicBoolean(false);
    private String mCurrentTimeStr = "";
    private String mCurrentMillisStr = "";

    private int mThemeMode = 0;
    private boolean mIsNightMode = false;

    // 优化: 复用通知 Builder，只更新变化的部分
    private Notification mCachedNotification;
    private String mLastTimeStr = "";
    private String mLastMillisStr = "";

    @Override
    public void onCreate() {
        super.onCreate();
        sIsRunning = true;
        mHandler = new Handler(Looper.getMainLooper());
        mPrefs = getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);

        mLiveUpdateManager = new LiveUpdateManager(this);

        loadPreferences();
        mNotificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);

        startForeground(NOTIFICATION_ID, createNotification());
        startClock();
        syncTime();

        log("FloatTimeService started, LiveUpdate supported: " + mLiveUpdateManager.isSupported());
    }

    private void loadPreferences() {
        mThemeMode = mPrefs.getInt(KEY_THEME_MODE, 0);
        mOffsetMs = mPrefs.getLong(KEY_OFFSET_MS, 0);
        mTimeSource = mPrefs.getString(KEY_TIME_SOURCE, "taobao");
        mIsNightMode = calcNightMode(mThemeMode);
    }

    // 优化: 抽取夜间模式计算逻辑，多处复用
    static boolean calcNightMode(int themeMode) {
        if (themeMode == 0) {
            int hour = java.util.Calendar.getInstance().get(java.util.Calendar.HOUR_OF_DAY);
            return hour >= 19 || hour < 7;
        }
        return themeMode == 2;
    }

    /**
     * 创建通知 (首次 + 每次需要重新构建时)
     */
    private Notification createNotification() {
        try {
            String timeStr = mCurrentTimeStr.isEmpty() ? "--:--:--" : mCurrentTimeStr;
            String millisStr = mCurrentMillisStr.isEmpty() ? ".000" : "." + mCurrentMillisStr;

            Notification notification = mLiveUpdateManager.createClockNotification(
                    this, timeStr, millisStr, getSourceDisplayName(), mIsNightMode
            );
            mCachedNotification = notification;
            mLastTimeStr = timeStr;
            mLastMillisStr = millisStr;
            return notification;
        } catch (Exception e) {
            log("createNotification error: " + e.getMessage());
            return createFallbackNotification();
        }
    }

    private Notification createFallbackNotification() {
        return new NotificationCompat.Builder(this, LiveUpdateManager.CHANNEL_ID)
                .setSmallIcon(android.R.drawable.ic_lock_idle_alarm)
                .setContentTitle("悬浮时间")
                .setContentText("服务运行中")
                .setOngoing(true)
                .build();
    }

    private void startClock() {
        mTimeRunnable = new Runnable() {
            @Override
            public void run() {
                updateTime();
                if (mHandler != null) {
                    mHandler.postDelayed(this, UPDATE_INTERVAL_MS);
                }
            }
        };
        mHandler.post(mTimeRunnable);
    }

    private void updateTime() {
        try {
            long now = System.currentTimeMillis() + mOffsetMs;
            Date date = new Date(now);

            String timeStr;
            synchronized (TIME_FORMAT) {
                timeStr = TIME_FORMAT.format(date);
            }

            mCurrentTimeStr = timeStr;
            mCurrentMillisStr = String.format(Locale.getDefault(), "%03d", now % 1000);

            // 主题切换检查 (优化: 用缓存的小时值,不再每次创建 Calendar)
            boolean newNight = calcNightMode(mThemeMode);
            if (newNight != mIsNightMode) {
                mIsNightMode = newNight;
                log("Theme switched: " + (mIsNightMode ? "dark" : "light"));
            }

            updateNotification();

        } catch (Exception e) {
            log("updateTime error: " + e.getMessage());
        }
    }

    /**
     * 更新通知
     * 优化: 直接复用 LiveUpdateManager.updateClock() 通知
     */
    private void updateNotification() {
        if (mNotificationManager == null) return;
        try {
            String timeStr = mCurrentTimeStr.isEmpty() ? "--:--:--" : mCurrentTimeStr;
            String millisStr = mCurrentMillisStr.isEmpty() ? ".000" : "." + mCurrentMillisStr;

            mLiveUpdateManager.updateClock(
                    this, timeStr, millisStr,
                    getSourceDisplayName(), mIsNightMode
            );
        } catch (Exception e) {
            log("updateNotification error: " + e.getMessage());
        }
    }

    private void syncTime() {
        if (!mIsSyncing.compareAndSet(false, true)) return;

        if ("local".equals(mTimeSource)) {
            mOffsetMs = 0;
            saveOffset();
            mLiveUpdateManager.showTimeSyncSuccess("local", 0);
            log("Local time mode, offset=0");
            mIsSyncing.set(false);
            return;
        }

        log("Syncing time from: " + mTimeSource);
        mLiveUpdateManager.showTimeSyncing(mTimeSource);

        new Thread(() -> {
            HttpURLConnection conn = null;
            try {
                String url = "taobao".equals(mTimeSource)
                        ? "https://api.m.taobao.com/rest/api3.do?api=mtop.common.getTimestamp"
                        : "https://api.meituan.com/nationalTimestamp";

                long localBefore = System.currentTimeMillis();

                conn = (HttpURLConnection) new URL(url).openConnection();
                conn.setConnectTimeout(CONNECT_TIMEOUT_MS);
                conn.setReadTimeout(READ_TIMEOUT_MS);
                conn.setRequestMethod("GET");
                conn.setRequestProperty("Accept", "application/json");

                int responseCode = conn.getResponseCode();
                if (responseCode == 200) {
                    java.io.BufferedReader reader = new java.io.BufferedReader(
                            new java.io.InputStreamReader(conn.getInputStream()));
                    StringBuilder response = new StringBuilder();
                    String line;
                    while ((line = reader.readLine()) != null) {
                        response.append(line);
                    }
                    reader.close();

                    long localAfter = System.currentTimeMillis();
                    long localMid = (localBefore + localAfter) / 2;
                    long serverTime = parseServerTime(response.toString());

                    if (serverTime > 0) {
                        mOffsetMs = serverTime - localMid;
                        saveOffset();
                        mLiveUpdateManager.showTimeSyncSuccess(mTimeSource, mOffsetMs);
                        log("Time synced: offset=" + mOffsetMs + "ms");
                    } else {
                        mLiveUpdateManager.showTimeSyncFailed(mTimeSource);
                        log("Failed to parse server time");
                    }
                } else {
                    mLiveUpdateManager.showTimeSyncFailed(mTimeSource);
                    log("HTTP error: " + responseCode);
                }
            } catch (Exception e) {
                mLiveUpdateManager.showTimeSyncFailed(mTimeSource);
                log("syncTime error: " + e.getMessage());
            } finally {
                if (conn != null) conn.disconnect();
                mIsSyncing.set(false);
            }
        }).start();
    }

    private long parseServerTime(String response) {
        try {
            JSONObject json = new JSONObject(response);
            long timestamp = 0;

            if (json.has("data")) {
                JSONObject data = json.getJSONObject("data");
                if (data.has("t")) {
                    timestamp = Long.parseLong(data.getString("t"));
                }
            }

            if (timestamp == 0 && json.has("t")) {
                timestamp = json.getLong("t");
            }

            if (timestamp == 0 && json.has("timestamp")) {
                timestamp = json.getLong("timestamp");
            }

            if (timestamp > 0) {
                return timestamp < 10000000000L ? timestamp * 1000 : timestamp;
            }
        } catch (Exception e) {
            log("parseServerTime error: " + e.getMessage());
        }
        return 0;
    }

    // 优化: SharedPreferences 写入节流 (最多每 5 秒写一次)
    private long mLastSaveTime = 0;

    private void saveOffset() {
        long now = System.currentTimeMillis();
        if (now - mLastSaveTime < 5000) return;
        mLastSaveTime = now;
        mPrefs.edit().putLong(KEY_OFFSET_MS, mOffsetMs).apply();
    }

    private String getSourceDisplayName() {
        switch (mTimeSource) {
            case "taobao": return "淘宝时间";
            case "meituan": return "美团时间";
            default: return "本地时间";
        }
    }

    private void log(String message) {
        Log.d(TAG, message);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            String action = intent.getAction();
            if ("STOP".equals(action)) {
                stopSelf();
            }
            // 修复: 处理主题切换 action (之前缺失)
            else if ("CHANGE_THEME".equals(action)) {
                int mode = intent.getIntExtra("mode", 0);
                mThemeMode = mode;
                mPrefs.edit().putInt(KEY_THEME_MODE, mode).apply();
                mIsNightMode = calcNightMode(mode);
                log("Theme changed to: " + mode + ", night=" + mIsNightMode);
            }
        }
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        log("onDestroy called");
        sIsRunning = false;

        if (mLiveUpdateManager != null) {
            mLiveUpdateManager.clearAll();
        }

        if (mHandler != null && mTimeRunnable != null) {
            mHandler.removeCallbacks(mTimeRunnable);
        }

        // 修复: 退出前保存偏移量
        mPrefs.edit().putLong(KEY_OFFSET_MS, mOffsetMs).apply();

        super.onDestroy();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/main/java/com/floattime/app/FloatTimeService.kt
package com.floattime.app

import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.util.Log
import androidx.core.app.NotificationCompat
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * 悬浮时间前台服务 (v3 - Live Island 集成版)
 *
 * 相比 v2 的改动:
 * - ✅ 新增: WindowAreaManager - 真正的 AndroidX Window Extensions API 接入 Live Island
 * - ✅ 新增: ClockTileService - 系统快捷设置磁贴入口
 * - ✅ 新增: NotificationLiveUpdate - Android 16 Notification Live Updates (ProgressStyle)
 * - ✅ 新增: HyperIslandManager - MIUI HyperOS 焦点通知增强
 * - ✅ 新增: ExtensionVersion 检测 - 运行时判断设备 Live Island 能力
 * - ✅ 提升: compileSdk/targetSdk 34 → 36, minSdk 27 → 30
 * - ✅ 移除: 原 SuperIslandManager (JSON 注入方案，替换为真正的 Extensions API)
 */
class FloatTimeService : Service() {

    companion object {
        private const val TAG = "FloatTimeService"
        private const val PREFS_NAME = "FloatTimePrefs"
        private const val KEY_THEME_MODE = "theme_mode"
        private const val KEY_OFFSET_MS = "offset_ms"
        private const val KEY_TIME_SOURCE = "time_source"

        private const val NOTIFICATION_ID = 20240320
        private const val UPDATE_INTERVAL_MS = 200L
        private const val CONNECT_TIMEOUT_MS = 8000L
        private const val READ_TIMEOUT_MS = 8000L
        private const val SAVE_THROTTLE_MS = 5000L
        private const val NOTIFICATION_UPDATE_MIN_INTERVAL_MS = 500L  // 降低更新间隔
        private const val NOTIFICATION_PERSISTENCE_CHECK_INTERVAL_MS = 3000L

        // ThreadLocal SimpleDateFormat: 每线程独立实例，零锁竞争
        private val TIME_FORMATTER = ThreadLocal.withInitial {
            SimpleDateFormat("HH:mm:ss", Locale.getDefault())
        }

        // 毫秒格式化 (独立，避免重复创建)
        private val MS_FORMAT = ThreadLocal.withInitial {
            SimpleDateFormat("SSS", Locale.US)  // Locale.US 保证3位数字
        }

        @Volatile
        private var sIsRunning = false

        fun isRunning(): Boolean = sIsRunning

        fun calcNightMode(themeMode: Int): Boolean = when (themeMode) {
            0 -> {
                val hour = java.util.Calendar.getInstance().get(java.util.Calendar.HOUR_OF_DAY)
                hour >= 19 || hour < 7
            }
            2 -> true
            else -> false
        }
    }

    private lateinit var notifMgr: NotificationManager
    private lateinit var islandManager: IslandManager
    private lateinit var prefs: SharedPreferences
    private lateinit var handler: Handler

    private var timeSource: String = "taobao"
    private var offsetMs: Long = 0
    private val isSyncing = AtomicBoolean(false)
    private var currentTimeStr: String = ""
    private var currentMillisStr: String = ""
    private var themeMode: Int = 0
    private var isNightMode: Boolean = false
    private var cachedTimezoneAbbr: String = ""
    private var lastSaveTime: Long = 0
    private var lastNotificationUpdateTime: Long = 0
    private var timeRunnable: Runnable? = null
    private var persistenceCheckRunnable: Runnable? = null

    private val networkReceiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context?, intent: Intent?) {
            if (intent?.action == ConnectivityManager.CONNECTIVITY_ACTION) {
                if (isNetworkAvailable()) {
                    Log.d(TAG, "Network restored, resyncing time...")
                    syncTime()
                }
            }
        }
    }

    // OkHttp 单例客户端（连接池，复用连接）
    private val httpClient: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(CONNECT_TIMEOUT_MS, TimeUnit.MILLISECONDS)
        .readTimeout(READ_TIMEOUT_MS, TimeUnit.MILLISECONDS)
        .retryOnConnectionFailure(true)
        .build()

    // ExecutorService 单例（单线程，避免并发问题）
    private val syncExecutor: ExecutorService = Executors.newSingleThreadExecutor { r ->
        Thread(r, "FloatTimeSync").apply { isDaemon = true }
    }

    override fun onCreate() {
        super.onCreate()
        sIsRunning = true
        handler = Handler(Looper.getMainLooper())
        prefs = getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

        // ✅ 核心改动: IslandManager 替换原 SuperIslandManager
        // 统一管理: Live Island (AndroidX) + 焦点通知 (MIUI) + 通知栏 (兼容)
        islandManager = IslandManager(this)

        loadPreferences()
        notifMgr = getSystemService(NOTIFICATION_SERVICE) as NotificationManager

        val filter = android.content.IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            registerReceiver(networkReceiver, filter, RECEIVER_NOT_EXPORTED)
        } else {
            registerReceiver(networkReceiver, filter)
        }

        startClock()

        handler.postDelayed({
            try {
                startForeground(NOTIFICATION_ID, createNotification())
                // ✅ 通知启动后，同时激活 Live Island
                islandManager.show(currentTimeStr, currentMillisStr, sourceDisplayName, isNightMode)
                Log.d(TAG, "Foreground service + Island started")
            } catch (e: Exception) {
                Log.e(TAG, "startForeground error: ${e.message}")
            }
        }, 100)

        ensureNotificationPersistent()
        syncTime()

        Log.d(TAG, "FloatTimeService started | Island: ${islandManager.describe()}")
    }

    private fun loadPreferences() {
        themeMode = prefs.getInt(KEY_THEME_MODE, 0)
        offsetMs = prefs.getLong(KEY_OFFSET_MS, 0)
        timeSource = prefs.getString(KEY_TIME_SOURCE, "taobao") ?: "taobao"
        isNightMode = calcNightMode(themeMode)
        cachedTimezoneAbbr = getTimezoneAbbr()
    }

    private fun getTimezoneAbbr(): String {
        return try {
            val tz = java.util.TimeZone.getDefault()
            tz.getDisplayName(isNightMode, java.util.TimeZone.SHORT, Locale.getDefault())
        } catch (_: Exception) {
            ""
        }
    }

    private fun createNotification(): Notification {
        val timeStr = currentTimeStr.ifEmpty { "--:--:--" }
        val millisStr = if (currentMillisStr.isEmpty()) ".000" else ".$currentMillisStr"

        return try {
            islandManager.createNotification(timeStr, millisStr, sourceDisplayName, isNightMode)
        } catch (e: Exception) {
            Log.e(TAG, "createNotification error: ${e.message}")
            createFallbackNotification()
        }
    }

    private fun createFallbackNotification(): Notification =
        NotificationCompat.Builder(this, IslandManager.CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_lock_idle_alarm)
            .setContentTitle("悬浮时间")
            .setContentText("服务运行中")
            .setOngoing(true)
            .build()

    private fun startClock() {
        timeRunnable = object : Runnable {
            override fun run() {
                updateTime()
                handler.postDelayed(this, UPDATE_INTERVAL_MS)
            }
        }.also { handler.post(it) }
    }

    private fun updateTime() {
        try {
            val now = System.currentTimeMillis() + offsetMs

            currentTimeStr = TIME_FORMATTER.get().format(Date(now))
            currentMillisStr = MS_FORMAT.get().format(Date(now))

            val newNight = calcNightMode(themeMode)
            if (newNight != isNightMode) {
                isNightMode = newNight
                cachedTimezoneAbbr = getTimezoneAbbr()
                Log.d(TAG, "Theme switched: ${if (isNightMode) "dark" else "light"}")
                // ✅ 主题切换时重建 Live Island
                islandManager.onThemeChanged(isNightMode)
            }

            updateNotification()
        } catch (e: Exception) {
            Log.e(TAG, "updateTime error: ${e.message}")
        }
    }

    private fun updateNotification() {
        if (!::notifMgr.isInitialized) return

        val now = System.currentTimeMillis()
        if (now - lastNotificationUpdateTime < NOTIFICATION_UPDATE_MIN_INTERVAL_MS) {
            return
        }
        lastNotificationUpdateTime = now

        try {
            val timeStr = currentTimeStr.ifEmpty { "--:--:--" }
            val millisStr = if (currentMillisStr.isEmpty()) ".000" else ".$currentMillisStr"

            val notification = islandManager.createNotification(timeStr, millisStr, sourceDisplayName, isNightMode)
            notifMgr.notify(NOTIFICATION_ID, notification)

            // ✅ 同步更新 Live Island
            islandManager.update(timeStr, millisStr, sourceDisplayName, isNightMode)
        } catch (e: Exception) {
            Log.e(TAG, "updateNotification error: ${e.message}")
        }
    }

    private fun ensureNotificationPersistent() {
        persistenceCheckRunnable = object : Runnable {
            override fun run() {
                if (sIsRunning && ::notifMgr.isInitialized) {
                    try {
                        val timeStr = currentTimeStr.ifEmpty { "--:--:--" }
                        val millisStr = if (currentMillisStr.isEmpty()) ".000" else ".$currentMillisStr"
                        val notification = islandManager.createNotification(
                            timeStr, millisStr, sourceDisplayName, isNightMode
                        )
                        startForeground(NOTIFICATION_ID, notification)
                        islandManager.show(timeStr, millisStr, sourceDisplayName, isNightMode)
                    } catch (e: Exception) {
                        Log.e(TAG, "ensureNotificationPersistent error: ${e.message}")
                    }
                }
                if (sIsRunning) {
                    handler.postDelayed(this, NOTIFICATION_PERSISTENCE_CHECK_INTERVAL_MS)
                }
            }
        }.also { handler.postDelayed(it, NOTIFICATION_PERSISTENCE_CHECK_INTERVAL_MS) }
    }

    private fun isNetworkAvailable(): Boolean {
        val cm = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val network = cm.activeNetwork ?: return false
        val capabilities = cm.getNetworkCapabilities(network) ?: return false
        return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
    }

    private fun syncTime() {
        if (!isSyncing.compareAndSet(false, true)) return

        if (timeSource == "local") {
            offsetMs = 0
            saveOffset()
            islandManager.showSyncResult(true, "local", 0)
            Log.d(TAG, "Local time mode, offset=0")
            isSyncing.set(false)
            return
        }

        Log.d(TAG, "Syncing time from: $timeSource")
        islandManager.showSyncStatus(true, timeSource)

        syncExecutor.submit {
            try {
                doSync()
            } finally {
                isSyncing.set(false)
            }
        }
    }

    private fun doSync() {
        val url = if (timeSource == "taobao")
            "https://api.m.taobao.com/rest/api3.do?api=mtop.common.getTimestamp"
        else
            "https://api.meituan.com/nationalTimestamp"

        val localBefore = System.currentTimeMillis()

        val request = Request.Builder()
            .url(url)
            .header("User-Agent", "FloatTime/" + getVersion(this))
            .header("Accept", "application/json")
            .build()

        try {
            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    postSyncFailed()
                    return
                }

                val responseBody = response.body?.string() ?: run {
                    postSyncFailed()
                    return
                }

                val localAfter = System.currentTimeMillis()
                val localMid = (localBefore + localAfter) / 2
                val serverTime = parseServerTime(responseBody)

                if (serverTime > 0) {
                    offsetMs = serverTime - localMid
                    saveOffset()
                    postSyncSuccess()
                    Log.d(TAG, "Time synced: offset=$offsetMs ms")
                } else {
                    postSyncFailed()
                    Log.d(TAG, "Failed to parse server time")
                }
            }
        } catch (e: IOException) {
            postSyncFailed()
            Log.e(TAG, "syncTime error: ${e.message}")
        }
    }

    private fun postSyncSuccess() {
        handler.post {
            try {
                islandManager.showSyncResult(true, timeSource, offsetMs)
            } catch (_: Exception) {}
        }
    }

    private fun postSyncFailed() {
        handler.post {
            try {
                islandManager.showSyncResult(false, timeSource, 0)
            } catch (_: Exception) {}
        }
    }

    private fun parseServerTime(response: String): Long {
        try {
            val json = JSONObject(response)
            if (json.has("data")) {
                val data = json.getJSONObject("data")
                if (data.has("t")) {
                    val ts = data.getString("t").toLongOrNull() ?: 0
                    if (ts > 0) return if (ts < 10000000000L) ts * 1000 else ts
                }
            }
            if (json.has("t")) {
                val ts = json.getLong("t")
                if (ts > 0) return if (ts < 10000000000L) ts * 1000 else ts
            }
            if (json.has("timestamp")) {
                val ts = json.getLong("timestamp")
                if (ts > 0) return if (ts < 10000000000L) ts * 1000 else ts
            }
        } catch (e: Exception) {
            Log.e(TAG, "parseServerTime error: ${e.message}")
        }
        return 0
    }

    private fun saveOffset() {
        val now = System.currentTimeMillis()
        if (now - lastSaveTime < SAVE_THROTTLE_MS) return
        lastSaveTime = now
        prefs.edit().putLong(KEY_OFFSET_MS, offsetMs).apply()
    }

    private val sourceDisplayName: String
        get() = when (timeSource) {
            "taobao" -> "淘宝时间"
            "meituan" -> "美团时间"
            else -> "本地时间"
        }

    private fun getVersion(ctx: Context): String {
        return try {
            ctx.packageManager.getPackageInfo(ctx.packageName, 0).versionName ?: "1.0"
        } catch (_: Exception) {
            "1.0"
        }
    }

    private fun log(msg: String) = Log.d(TAG, msg)

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            "STOP" -> {
                stopForeground(STOP_FOREGROUND_REMOVE)
                islandManager.hide()
                stopSelf()
            }
            "CHANGE_THEME" -> {
                val mode = intent.getIntExtra("mode", 0)
                themeMode = mode
                prefs.edit().putInt(KEY_THEME_MODE, mode).apply()
                isNightMode = calcNightMode(mode)
                cachedTimezoneAbbr = getTimezoneAbbr()
                islandManager.onThemeChanged(isNightMode)
                Log.d(TAG, "Theme changed to: $mode, night=$isNightMode")
            }
        }
        return START_STICKY
    }

    override fun onDestroy() {
        log("onDestroy called")
        sIsRunning = false

        try {
            unregisterReceiver(networkReceiver)
        } catch (_: Exception) {}

        if (::islandManager.isInitialized) {
            islandManager.destroy()
        }

        timeRunnable?.let { handler.removeCallbacks(it) }
        persistenceCheckRunnable?.let { handler.removeCallbacks(it) }
        prefs.edit().putLong(KEY_OFFSET_MS, offsetMs).apply()

        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? = null
}
